src/platform/OpenThread/GenericNetworkCommissioningThreadDriver.cpp

```diff
@@ -55,17 +65,41 @@ CHIP_ERROR GenericThreadDriver::Shutdown()
 
 CHIP_ERROR GenericThreadDriver::CommitConfiguration()
 {
-    // Note: on AttachToThreadNetwork OpenThread will persist the networks on its own,
-    // we don't have much to do for saving the networks (see Init() above,
-    // we just load the saved dataset from ot instance.)
-    mSavedNetwork = mStagingNetwork;
-    return CHIP_NO_ERROR;
+    // OpenThread persists the network configuration on AttachToThreadNetwork, so simply remove
+    // the backup, so that it cannot be restored. If no backup could be found, it means that the
+    // configuration has not been modified since the fail-safe was armed, so return with no error.
+
+    CHIP_ERROR error = KeyValueStoreMgr().Delete(DefaultStorageKeyAllocator::FailSafeNetworkConfig());
+
+    return error == CHIP_ERROR_PERSISTED_STORAGE_VALUE_NOT_FOUND ? CHIP_NO_ERROR : error;
 }
 
 CHIP_ERROR GenericThreadDriver::RevertConfiguration()
 {
-    mStagingNetwork = mSavedNetwork;
-    return CHIP_NO_ERROR;
+    uint8_t datasetBytes[Thread::kSizeOperationalDataset];
+    size_t datasetLength;
+
+    CHIP_ERROR error = KeyValueStoreMgr().Get(DefaultStorageKeyAllocator::FailSafeNetworkConfig(), datasetBytes,
+                                              sizeof(datasetBytes), &datasetLength);
+
+    // If no backup could be found, it means that the network configuration has not been modified
+    // since the fail-safe was armed, so return with no error.
+    ReturnErrorCodeIf(error == CHIP_ERROR_PERSISTED_STORAGE_VALUE_NOT_FOUND, CHIP_NO_ERROR);
+    ReturnErrorOnFailure(error);
+
+    KeyValueStoreMgr().Delete(DefaultStorageKeyAllocator::FailSafeNetworkConfig());
```

Comment:
If we crash or get powered down after this call but before we manage to persist the restored value (presumably under `AttachToThreadNetwork`?) we will end up in a bad state, no?

It seems like we should only do this deletion after we know the commit to storage of the restored value has succeeded.

Reply:
Right, makes sense.